### ISSUE TITLE
Feature Request: Sampled softmax loss

### ISSUE THREAD

**Issue by mruberry** [pytorch/pytorch#38185]
Analogous to TensorFlow's https://www.tensorflow.org/api_docs/python/tf/nn/sampled_softmax_loss. 

This was first proposed in #3718.

**Comment by vadimkantorov** [comment]
Related feature request: https://github.com/pytorch/pytorch/issues/634
